### DISCUSSION TITLE
louvain_communities runs endlessly

### DISCUSSION THREAD

**Discussion by omri-al** [networkx/networkx#7307] [Q&A]
Hi,
I have a program that creates thousands of graphs using networkx, and for each one of them, I extract the communities using louvain_communities algorithm.
I have discovered that for some of these graphs, which are not tha big at all (around 100-200 nodes and edges) the algorithm just get stuck and never ends (runs for days instead of seconds...).
Any idea what can lead to that?
And more important: how can I prevent it?
I tried using different "thresholds" as parameter to the function, but couldn't find a solution. I would expect to have something like "max_iteration" to prevent it from running too long.
Any advice?
thanks

**Comment by dschult** [comment]
I believe the argument you are looking for is `max_level`. 
That is the term to describe how many "iterations" or "levels" you will compute.

**Reply by eriknw** [reply]
`max_level` for louvain was just added in #6909.

**Reply by rossbar** [reply]
As mentioned, the `max_level` arg has been added to allow users to limit this explicitly and will be available in NetworkX 3.3.

Re: the underlying cause of the issue - I suspect it is due to floating point comparison. Others have had similar issues, see e.g. #6751 .

**Reply by dschult** [reply]
In 3.2 and earlier you can use the [`nx.louvain_partitions`](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.community.louvain.louvain_partitions.html)  generator which yields the algorithm results after each step of the process.

**Reply by omri-al** [reply]
@dschult but wouldn't [nx.louvain_partitions] stuck on the same level and I will remain with the same issue?

**Reply by dschult** [reply]
I was thinking this would solve your issue (when using 3.2 and before) because you could control when to stop the iterator. Something like:
```python
for i, parts in nx.louvain_partitions(...):
    if i > max_its:
        break
```
